trieste/acquisition/function/greedy_batch.py

```diff
@@ -351,3 +355,268 @@ def __call__(self, x: TensorType) -> TensorType:
+            if datasets is None:
+                datasets = fantasized_data
+            else:
+                datasets = {tag: data + fantasized_data[tag] for tag, data in datasets.items()}
+        return self._builder.prepare_acquisition_function(models, datasets)
+
+    def _generate_fantasized_data(
+        self, model: ProbabilisticModel, pending_points: TensorType
+    ) -> Dataset:
+        if self._fantasize_method == "KB":
+            fantasized_obs, _ = model.predict(pending_points)
+        else:
+            fantasized_obs = model.sample(pending_points, num_samples=1)  # TODO check dim
+
+        return Dataset(pending_points, fantasized_obs)
+
+
+def _fantasize_model(
+    model: ProbabilisticModel, fantasized_data: Dataset
+) -> _fantasized_model | ModelStack:
+    if isinstance(model, ModelStack):
+        observations = tf.split(fantasized_data.observations, model._event_sizes, axis=-1)
+        fmods = []
+        for mod, obs, event_size in zip(model._models, observations, model._event_sizes):
+            fmods.append(
+                (
+                    _fantasized_model(mod, Dataset(fantasized_data.query_points, obs)),
+                    event_size,
+                )
+            )
+        return ModelStack(*fmods)
+    else:
+        return _fantasized_model(model, fantasized_data)
+
+
+class _fantasized_model(ProbabilisticModel):
+    """
+    Creates a new model from an existing one and additional data.
+    This new model posterior is conditioned on both current model data and the additional one.
+    """
+
+    def __init__(self, model: ProbabilisticModel, fantasized_data: Dataset):
+        """
+        :param model: a model, must be of :class:`~FantasizedGPRModel`
+        :param fantasized_data: additional dataset to condition on
+        :raise NotImplementedError: If model is not of :class:`~FantasizedGPRModel`.
+        """
+
+        if not isinstance(model, FastUpdateModel):
+            raise NotImplementedError(
+                f"FantasizedAcquisitionFunction only works with FastUpdateModel, "
+                f"received "
+                f"{model.__repr__()}"
+            )
+
+        self._model = model
+        self._fantasized_data = fantasized_data
+
+    def predict(self, query_points: TensorType) -> tuple[TensorType, TensorType]:
+        """
+        This function wraps conditional_predict_f. It cannot directly call
+        conditional_predict_f, since it does not accept query_points with rank > 2.
+        We use map_fn to allow leading dimensions for query_points.
+
+        :param query_points: shape [...*, n, d]
+        :return: mean, shape [...*, ..., n, L] and cov, shape [...*, ..., L, n],
+            where ... are the leading dimensions of fantasized_data
+        """
+
+        def fun(qp: TensorType) -> tuple[TensorType, TensorType]:
+            return self._model.conditional_predict_f(qp, self._fantasized_data)
+
+        return _broadcast_predict(query_points, fun)
```

Comment:
this seems like it adds a lot of overhead... perhaps @uri-granta has an advice how to optimize it further?

Reply:
Yes, this is a bit unfortunate. But in terms of efficiency, I don't think it's as efficient as automatic broadcasting...